### ISSUE TITLE
ttnn.layer_norm low PCC for a specific input found in albert-v2-base

### ISSUE THREAD

**Issue by amalbasaTT** [tenstorrent/tt-metal#18345]
**Describe the bug**
ttnn.layer_norm has a low PCC (0.998992414852081) for a specific input found in albert-v2-base, which could be the cause of low accuracy in albert-v2 models. Inputs specifications are next:

input:

- values: [ln_25_input.csv](https://github.com/user-attachments/files/18985447/ln_25_input.csv)
- shape: (1, 9, 128)
- dtype: ttnn.bfloat16
- layout: ttnn.TILE_LAYOUT

weight:
- values: [ln_25_weight.csv](https://github.com/user-attachments/files/18985459/ln_25_weight.csv)
- shape (128)
- dtype: ttnn.bfloat16
- layout: ttnn.TILE_LAYOUT

bias:
- values: [ln_25_bias.csv](https://github.com/user-attachments/files/18985466/ln_25_bias.csv)
- shape (128)
- dtype: ttnn.bfloat16
- layout: ttnn.TILE_LAYOUT

**To Reproduce**
Run unit test (contains function for loading tensors from csv file):
```
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten

    
def load_tensor_from_csv(filename, shape):
    import pandas as pd
    df = pd.read_csv(filename)
    data = df.values
    tensor = torch.tensor(data)
    # Reshape the tensor to the original shape
    tensor = tensor.view(shape)
    return tensor


def run_layer_norm_tests(
    file_path,
    input_shape, 
    dtype, 
    dlayout,
    device,
):
    x = load_tensor_from_csv(file_path[0], input_shape[0]).to(torch.bfloat16)
    weight = load_tensor_from_csv(file_path[1], input_shape[1]).to(torch.bfloat16)
    bias = load_tensor_from_csv(file_path[2], input_shape[2]).to(torch.bfloat16)
    epsilon = 1e-12

    try:
        # get ref result
        ref_value = aten.native_layer_norm.default(x, [input_shape[0][-1]], weight, bias, epsilon)[0]
        
        tt_x = ttnn.from_torch(x, dtype=dtype[0], layout=dlayout[0], device=device)
        tt_weight = ttnn.from_torch(weight, dtype=dtype[0], layout=dlayout[0], device=device)
        tt_bias = ttnn.from_torch(bias, dtype=dtype[0], layout=dlayout[0], device=device)

        tt_result = ttnn.layer_norm(tt_x, epsilon=epsilon, weight=tt_weight, bias=tt_bias)
        tt_result = ttnn.to_torch(tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    assert_with_pcc(ref_value, tt_result, 0.999)


test_sweep_args = [
    (
        ["/home/ubuntu/pytorch2.0_ttnn/ln_25_input.csv", 
         "/home/ubuntu/pytorch2.0_ttnn/ln_25_weight.csv", 
         "/home/ubuntu/pytorch2.0_ttnn/ln_25_bias.csv"],
        [(1, 9, 128), (128), (128)],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
    ),
]


@pytest.mark.parametrize(
    "file_path, input_shape, dtype, dlayout",
    (test_sweep_args),
)
def test_layer_norm(file_path, input_shape, dtype, dlayout, device):
    run_layer_norm_tests(file_path, input_shape, dtype, dlayout, device)

```

**Comment by bbradelTT** [comment]
@amalbasaTT the default accuracy is not sufficient. You need to enable fp32_dest_acc_en.

Can you please try replacing the call to ttnn.layer_norm with the following and let me know if that fixes your issue? The pcc should be above 0.999
```
        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=False,
        )
        tt_result = ttnn.layer_norm(tt_x, epsilon=epsilon, weight=tt_weight, bias=tt_bias, compute_kernel_config=compute_kernel_config)
```

**Comment by amalbasaTT** [comment]
@bbradelTT PCC is now 0.999263340147414.

**Comment by bbradelTT** [comment]
@amalbasaTT are you okay with me closing this issue?

**Comment by bbradelTT** [comment]
One potential option would be that I could set that value as the default internally.

**Comment by bbradelTT** [comment]
Update:
It looks like Falcon 7B uses default values, and the tensors are large enough that using fp32 causes the op not to work. My change won't pass all post commit without changing various models or rewriting the op (which might be done at some point).

Also, compared to the other PCCs in the parent issue, 0.99899 is much better and probably not a source of the problem.

**Comment by bbradelTT** [comment]
@amalbasaTT would it be acceptable for you to use torch.float32 / ttnn.float32 tensors?

I might be able to conditionally set the config if inputs are float32.

**Comment by bbradelTT** [comment]
Update:
@amalbasaTT will provide more info about which options are acceptable in a couple of days.

**Comment by amalbasaTT** [comment]
As for now, including compute_kernels for ttnn_layer_norm ops increases overall model accuracy from 0.96 to 0.9886.

**Comment by ayerofieiev-tt** [comment]
@amalbasaTT lets explicitly set config to fp32 on our side. 
 

@bbradelTT op should work without requiring us to pass this config. We are aligned here, right?

 I understand that existing models work without passing this config and a change to the default behavior of the op will cause some of them to fail. Means changing the default behavior is a journey. 

I think we are aligned that every op has to take that journey at some point. The tradeoff is -  the longer Ie’s delay the more work we have to do. So I’d recommend to find xfn time to make it. Ultimately it is your decision.

**Comment by amalbasaTT** [comment]
Using kernel config reduces PCC of mixer_b16_224.goog_in21k_code to 0.0364.

**Comment by nemanjagrujic** [comment]
> PCC of mixer_b16_224.goog_in21k_code

@amalbasaTT What was previous PCC for mixer_b16_224.goog_in21k_code?

**Comment by amalbasaTT** [comment]
0.53

**Comment by bbradelTT** [comment]
@amalbasaTT The code is merged. Please test it out and if everything works, please close the issue.

**Comment by amalbasaTT** [comment]
-  ALBERT_V2_BASE
Previous accuracy: 0.69
Expected accuracy: 0.758
Accuracy now: 0.757
-  ALBERT_V2_LARGE
Previous accuracy:  0.236
Expected accuracy: 0.177
Accuracy now: 0.201
-  ALBERT_V2_XLARGE
Previous accuracy: 0.534
Expected accuracy: 0.547
Accuracy now: 0.592
- ALBERT_V2_XXLARGE
Previous accuracy: 0.186
Expected accuracy: 0.211
Accuracy now: 0.214
- mixer_b16_224.goog_in21k
Previous accuracy: 0.538
Expected accuracy: 0.928
Accuracy now: 0.036
Accuracy when casting input_tensor to float32 and casting layer_norm result back to bfloat16: 0.566


@bbradelTT  For some reason, your changes are drastically reducing PCC of mixer_b16_224.goog_in21k model. This particular issue is different from the one in mixer model (https://github.com/tenstorrent/tt-metal/issues/18548), but your changes make it even worse than it already is. Casting only input tensor to float32 and casting layer_norm result back to bfloat16 seem to slightly increase the PCC of mixer model.
@ayerofieiev-tt should we close this issue or not? And if not, should I include typcast of input tensor for all layer_norms on pytorch2.0_ttnn project?

**Comment by bbradelTT** [comment]
Please create a new issue for the mixer model with repro steps. That seems like it's running into a bug. We would investigate that separately.

**Comment by bbradelTT** [comment]
And repro steps might be to just use what's in https://github.com/tenstorrent/tt-metal/issues/18548

**Comment by nemanjagrujic** [comment]
@bbradelTT Can you send link to pull req for this change?

**Comment by amalbasaTT** [comment]
@nemanjagrujic https://github.com/tenstorrent/tt-metal/pull/19295

**Comment by bbradelTT** [comment]
@nemanjagrujic github includes the link to the commit in main and the link to the PR should be part of that. Look for the verified commit. E.g.

![Image](https://github.com/user-attachments/assets/24f8ffc6-f563-493a-a479-1669fe7d1f43)

**Comment by nemanjagrujic** [comment]
@bbradelTT OK, I see it now
